Source/WebKit/Shared/Extensions/_WKWebExtensionLocalization.mm

```diff
@@ -0,0 +1,372 @@
+        };
+    }
+
+    if (_uniqueIdentifier)
+        return mergeDictionaries(predefindedMessage, @{ predefinedMessageExtensionID: @{ _WKLocalizationDictionaryMessageKey: _uniqueIdentifier } });
+    return predefindedMessage;
+}
+
+- (NSString *)_stringByReplacingNamedPlaceholdersInString:(NSString *)string withNamedPlaceholders:(PlaceholderDictionary *)namedPlaceholders
+{
+    NSRegularExpression *namedPlaceholderExpression = [NSRegularExpression regularExpressionWithPattern:@"(?:[^$]|^)(\\$([A-Za-z0-9_@]+)\\$)" options:0 error:nil];
+    NSTextCheckingResult *namedPlaceholderMatch = [namedPlaceholderExpression firstMatchInString:string options:0 range:NSMakeRange(0, string.length)];
+    while (namedPlaceholderMatch) {
+        NSString *placeholderKey = [string substringWithRange:[namedPlaceholderMatch rangeAtIndex:2]].lowercaseString;
+        NSString *replacement = objectForKey<NSString>(objectForKey<NSDictionary>(namedPlaceholders, placeholderKey), placeholderDictionaryContentKey) ?: @"";
+
+        NSString *stringToReplace = [string substringWithRange:[namedPlaceholderMatch rangeAtIndex:1]];
+
+        string = [string stringByReplacingOccurrencesOfString:stringToReplace withString:replacement options:NSCaseInsensitiveSearch range:NSMakeRange(0, string.length)];
+
+        namedPlaceholderMatch = [namedPlaceholderExpression firstMatchInString:string options:0 range:NSMakeRange(0, string.length)];
+    }
+
+    return string;
+}
+
+- (NSString *)_stringByReplacingPositionalPlaceholdersInString:(NSString *)string withValues:(NSArray<NSString *> *)placeholderValues
+{
+    NSInteger placeholderCount = placeholderValues.count;
+
+    NSRegularExpression *positionalPlaceholderExpression = [NSRegularExpression regularExpressionWithPattern:@"(?:[^$]|^)(\\$([0-9]))" options:0 error:nil];
+    NSTextCheckingResult *positionalPlaceholderMatch = [positionalPlaceholderExpression firstMatchInString:string options:0 range:NSMakeRange(0, string.length)];
+
+    while (positionalPlaceholderMatch) {
+        NSInteger position = [string substringWithRange:[positionalPlaceholderMatch rangeAtIndex:2]].integerValue;
+
+        NSString *replacement;
+        if (position > 0 && position <= placeholderCount)
+            replacement = placeholderValues[position - 1];
+        else
+            replacement = @"";
+
+        NSString *stringToReplace = [string substringWithRange:[positionalPlaceholderMatch rangeAtIndex:1]];
+
+        string = [string stringByReplacingOccurrencesOfString:stringToReplace withString:replacement options:NSCaseInsensitiveSearch range:NSMakeRange(0, string.length)];
+        positionalPlaceholderMatch = [positionalPlaceholderExpression firstMatchInString:string options:0 range:NSMakeRange(0, string.length)];
+    }
+
+    return string;
+}
+
+// MARK: NSSecureCoding
+
++ (BOOL)supportsSecureCoding
+{
+    return YES;
+}
```

Comment:
Move to the top, outside the `#if ENABLE(WK_WEB_EXTENSIONS)` but just inside the `@implementation`.